### PR TITLE
feat: add `MouseMove` autocmd event

### PR DESCRIPTION
**Pull request by willothy**
Problem: Some plugins, mainly statusline / tabline plugins, are implementing hover handlers / drag handlers which require access to the `<MouseMove>` event. However, mappings between these plugins can conflict.

Solution: This module allows plugins / user functions to individually subscribe to mouse move events without handling the event logic. 

<strike>On each event, the global handler iterates over subscribed callbacks, passing the result of `vim.fn.getmousepos()`. Essentially, it does what plugins that implement hover / drag functionality are already doing, but in a global interface that won't cause any conflicts between plugins.</strike>

I know most users don't use the mouse, but this would be an entirely optional feature for those who do, and who don't mind the overhead. This code isn't intended to be final, but more of a conversation starter. I included the setup function for now, but it would probably make more sense to use a global option and handle setup automatically for users who enable it. Not sure how to add options though, this is my first PR here :)

### CODE REVIEW

File: runtime/lua/vim/mouse.lua [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+---  - wincol (number)
+---  - line (number)
+---  - column (number)
+---@param handler fun(mouse: MouseInfo)
+function M.subscribe(handler)
+  table.insert(subscribers, handler)
+end
+
+function M.setup()
+  if vim.version().minor > 8 and vim.o.mousemoveevent then
```
Comment (justinmk) [inline runtime/lua/vim/mouse.lua:49]: checking only `minor` is the wrong way to check nvim version. could use `has('nvim-0.8')` instead.

Comment (clason) [reply runtime/lua/vim/mouse.lua:49]: We shouldn't check the version here in any case, as we _know_ what version we are running under (0.10.0 or higher).

**Comment by justinmk** [comment]
> This code isn't intended to be final, but more of a conversation starter.

Nicely done, the "problem" statement especially helps. 

But I don't fully understand what is being solved yet. "mappings conflict" and "plugins already do this"... what is missing from the current event system. We already have too many "event interfaces" https://github.com/neovim/neovim/issues/19156 , don't want to add another one.

File: runtime/lua/vim/mouse.lua (hunk 1 above)
Comment (willothy) [reply runtime/lua/vim/mouse.lua:49]: Definitely was one of the things I wasn't sure about. And my bad, should've marked this as a draft.

**Comment by willothy** [comment]
> But I don't fully understand what is being solved yet. "mappings conflict" and "plugins already do this"... what is missing from the current event system? We already have too many "event interfaces" #19156 , don't want to add another one.

Essentially, the issue is that whenever a user or plugin uses `vim.keymap.set` to map `<MouseMove>`, other `<MouseMove>` mappings no longer work. Additionally, even if multiple mappings could be registered for mouse events, all handlers would be calling vim.fn.mousepos() (a secondary concern but still worth noting). 
If several plugins wanted to handle mouse move events, they would need to use a common list of handlers and a single mapping that calls all of them. This makes that idea a public nvim interface instead of relying on plugins to coordinate and use a shared global or something of that nature.

> How would you change the existing event system to better support the use-case?

I went with this approach because I'm far more familiar with Lua than I am with Neovim's event loop, but I think that making certain input events act like broadcast channels / autocmds would make this change unnecessary, since multiple sources could receive the events. Not sure if that's possible, but I'd definitely be interested in contributing any other change that'd make this work.

**Comment by willothy** [comment]
> Ok so this is more like a request for key mappings to support multiple "handlers".
> 
> Another way to achieve this specific case: add a `MouseMove` event?

Yep! And do you mean event as in Autocmd event? That would definitely work.

Also, after reading through #19156 I think it would make sense if this were handled by `vim.on` if that gets implemented/merged.

**Comment by willothy** [comment]
Ok, spent some time looking through the C code and got a MouseMoved autocmd working. It's pretty basic and doesn't work when dragging atm but it's a start, will update once I get a bit more familiar with the codebase :)

File: src/nvim/mouse.c [hunk 2]
```diff
@@ -246,6 +246,16 @@ static int get_fpos_of_mouse(pos_T *mpos)
   return IN_BUFFER;
 }
 
+void do_move_autocmd(int screenrow, int screencol) {
+	Dictionary data = ARRAY_DICT_INIT;
+	PUT(data, "screenrow", INTEGER_OBJ(screenrow));
+	PUT(data, "screencol", INTEGER_OBJ(screencol));
+	Object obj = DICTIONARY_OBJ(data);
+	apply_autocmds_group(EVENT_MOUSEMOVED, NULL, NULL, false, AUGROUP_ALL, NULL, NULL, &obj);
+	api_free_dictionary(data);
```
Comment (willothy) [inline src/nvim/mouse.c:255]: Could use input on this - I saw similar free calls used in other files, but not sure if this is actually the proper way to free these values or if freeing the Object and its underlying Dictionary will result in a double free.

File: src/nvim/mouse.c [hunk 3]
```diff
@@ -246,6 +246,16 @@ static int get_fpos_of_mouse(pos_T *mpos)
   return IN_BUFFER;
 }
 
+void do_move_autocmd(int screenrow, int screencol) {
+	Dictionary data = ARRAY_DICT_INIT;
+	PUT(data, "screenrow", INTEGER_OBJ(screenrow));
```
Comment (willothy) [inline src/nvim/mouse.c:251]: I don't think there's a way to call mousepos() from here, so passing the screenrow and screencol should do for now.

Comment (willothy) [reply src/nvim/mouse.c]: Currently getting exactly what I needed out of this and it appears to be working fine, so gonna mark as ready for review.

File: src/nvim/mouse.c [hunk 4]
```diff
@@ -245,6 +246,15 @@ static int get_fpos_of_mouse(pos_T *mpos)
   return IN_BUFFER;
 }
 
+void do_move_autocmd(int screenrow, int screencol) {
```
Comment (zeertzjq) [inline src/nvim/mouse.c:249]:
```suggestion
static void do_move_autocmd(int screenrow, int screencol) {
```

File: src/nvim/mouse.c [hunk 5]
```diff
@@ -245,6 +246,16 @@ static int get_fpos_of_mouse(pos_T *mpos)
...
 }
 
+/// Execute mouse move autocmd.
+static void do_move_autocmd(int screenrow, int screencol) {
+	Dictionary data = ARRAY_DICT_INIT;
+	PUT(data, "screenrow", INTEGER_OBJ(screenrow + 1));
+	PUT(data, "screencol", INTEGER_OBJ(screencol + 1));
+	Object obj = DICTIONARY_OBJ(data);
+	apply_autocmds_group(EVENT_MOUSEMOVED, NULL, NULL, false, AUGROUP_ALL, NULL, NULL, &obj);
+	api_free_dictionary(data);
```
Comment (willothy) [inline src/nvim/mouse.c:256]: Not sure if I should be calling `api_free_dictionary(data)` or `api_free_object(obj)` here

File: src/nvim/mouse.c [hunk 6]
```diff
@@ -245,6 +246,27 @@ static int get_fpos_of_mouse(pos_T *mpos)
   return IN_BUFFER;
 }
 
+/// Execute mouse move autocmd.
+static void do_move_autocmd(int screenrow, int screencol) {
```
Comment (justinmk) [inline src/nvim/mouse.c:250]:
```suggestion
static void do_mousemove_autocmd(int screenrow, int screencol) {
```

File: src/nvim/mouse.c (hunk 5 above)
Comment (justinmk) [reply src/nvim/mouse.c:267]: `api_free_object` , else you need to do `api_free_dictionary(data.dictionary)`

File: src/nvim/mouse.c [hunk 7]
```diff
@@ -466,6 +491,8 @@ bool do_mouse(oparg_T *oap, int c, int dir, long count, bool fixindent)
   // When dragging or button-up stay in the same window.
   if (!is_click) {
     jump_flags |= MOUSE_FOCUS | MOUSE_DID_MOVE;
+	// Execute the MouseMoved autocmd
+	do_move_autocmd(mouse_row, mouse_col);
```
Comment (justinmk) [inline src/nvim/mouse.c:495]: are these the correct places for emitting the event? Check the git log for when `<MouseMove>` was added to see hints for affected code.

need test coverage in `test/functional/ui/mouse_spec.lua`.

Comment (willothy) [reply src/nvim/mouse.c:495]: I believe this is correct for dragging, but I'll definitely look into it more. Could also be removed or used for a separate drag event if that would make more sense.

**Comment by willothy** [comment]
I'd like to provide some info about dragging, here are my ideas so far:
- Separate `MouseDrag` autocmd that provides a location and drag button(s)
- Include an "is_dragging" field on the table passed to `MouseMove` that either is nil or contains a string/number representing the mouse button(s) being pressed during the drag.
- Include an "is_dragging" field, but only as a boolean and make plugins handle click detection on their own.

**Comment by willothy** [comment]
My bad, renamed the branch and wasn't aware that would close the PR.

**Comment by willothy** [comment]
Anything else I need to get done for this? Or should I look into alternate implementations? I noticed that there's a `vim.on_key` function - Maybe something like `vim.on_mouse` or `vim.on_drag` could be used instead of an autocmd if that makes more sense.

**Review by zeertzjq — CHANGES_REQUESTED** [review]
I don't think it's really useful to add this event. This can be already be done in a safer and more flexible way using `vim.on_key()`:
```lua
local mouse_move = vim.keycode('<MouseMove>')
local drag_key = {
  [vim.keycode('<LeftDrag>')] = 'l',
  [vim.keycode('<RightDrag>')] = 'r',
  [vim.keycode('<MiddleDrag>')] = 'm',
}
vim.on_key(function(key)
  if key == mouse_move or drag_key[key] then
    vim.schedule_wrap(function(key)
      local data = vim.fn.getmousepos()
      data.dragging = drag_key[key]
      vim.api.nvim_exec_autocmds('User', { pattern = 'MouseMove', data = data })
    end)(key)
  end
end)
```

**Comment by willothy** [comment]
Oh, perfect! Thanks for the reviews anyways, had a good time learning a bit about the codebase. Hope I can contribute something more useful in the future :)

File: runtime/doc/autocmd.txt [hunk 8]
```diff
@@ -1131,6 +1131,33 @@ WinResized			After a window in the current tab page changed
...
+
+							*MouseMove*
+MouseMove			After the mouse is moved. Similar to the `<MouseMove>`
+				mapping.
+
+				Unlike the `<MouseMove>` mapping, it also triggers when
+				dragging.
+
+				|v:event| is set with information about the mouse position,
+				which can be used like this: >
```
[the remark this answers is not in the document]
Comment (willothy) [reply runtime/doc/autocmd.txt:1143]: Ah, I see. Would this mean that the data would only be accessible from Lua, and not Vimscript? I'll look into setting v:event for future reference. Thanks!

**Pull request closed without merging**